contracts/SeriesIndexer.sol

```diff
@@ -0,0 +1,448 @@
+        _checkMetadataAndTokenCID(metadata, tokenIDsMapCID);
+
+        // Mark series as existing
+        seriesDetails[seriesID].metadata = metadata;
+        seriesDetails[seriesID].contractTokenData = tokenIDsMapCID;
+
+        uint256[] memory artistIDs = _addArtistsToSeries(seriesID, artistAddrs);
+
+        emit SeriesIndexed(seriesID, artistIDs, metadata, tokenIDsMapCID);
+        return seriesID;
+    }
+
+    function _removeSeriesFromArtist(uint256 artistID, uint256 seriesID) internal {
+        uint256[] storage seriesIDs = artistIDSeriesIDs[artistID];
+        uint256 length = seriesIDs.length;
+        for (uint256 i = 0; i < length; i++) {
+            if (seriesIDs[i] == seriesID) {
+                seriesIDs[i] = seriesIDs[length - 1];
+                seriesIDs.pop();
+                break;
+            }
+        }
+        isArtistIDInSeries[seriesID][artistID] = false;
+    }
+
+    function _removeArtistFromSeries(uint256 seriesID, uint256 artistID) internal {
+        uint256[] storage artistIDs = seriesDetails[seriesID].artistIDs;
+        uint256 length = artistIDs.length;
+        // Remove artistID from series' artist list
+        for (uint256 i = 0; i < length; i++) {
+            if (artistIDs[i] == artistID) {
+                artistIDs[i] = artistIDs[length - 1];
+                artistIDs.pop();
+                break;
+            }
+        }
+
+        // Remove series from artist's series list
+        _removeSeriesFromArtist(artistID, seriesID);
+
+        emit SeriesUpdated(seriesID, artistIDs, seriesDetails[seriesID].metadata, seriesDetails[seriesID].contractTokenData);
+    }
+
+    // ------------------------
+    // External and Public Functions
+    // ------------------------
+
+    function addSeries(
+        string calldata metadata,
+        string calldata tokenIDsMapCID
+    ) external returns (uint256) {
+        address[] memory artistAddrs = new address[](1);
+        artistAddrs[0] = msg.sender;
+        return _createSeries(artistAddrs, metadata, tokenIDsMapCID);
+    }
+
+    function ownerAddSeries(
+        address[] calldata artistAddrs,
+        string calldata metadata,
+        string calldata tokenIDsMapCID
+    ) external onlyOwner returns (uint256) {
+        require(artistAddrs.length > 0, "No artists");
+        _checkArtistsNotRevoked(artistAddrs);
+        return _createSeries(artistAddrs, metadata, tokenIDsMapCID);
+    }
+
+    function ownerBatchAddSeries(
+        address[][] calldata artistsArray,
+        string[] calldata metadatas,
+        string[] calldata tokenIDsMapCIDs
+    ) external onlyOwner returns (uint256[] memory) {
+        uint256 length = artistsArray.length;
+        require(
+            length == metadatas.length &&
+            length == tokenIDsMapCIDs.length,
+            "Array length mismatch"
+        );
+        require(length > 0, "No series data");
+        require(length <= 50, "Batch size too large"); // Prevent DOS attacks
+
```

Comment:
What does this mean for prevent DOS attacks?

Reply:
that mean if someone try to put really large number of list, the contract will deny to service. It's not quite of an attack I think.

Reply:
the words is removed anyway

Reply:
So that could be out of gas since it could reach the block gas limit so i don't consider it as "denial of service"